testing/test_warnings.py

```diff
@@ -640,3 +641,117 @@ def pytest_configure():
+                "*could not load initial conftests*",
+            ]
+        )
+
+    def test_issue4445_import_plugin(self, testdir):
+        """#4445: Make sure the warning points to a reasonable location
+        See origin of _issue_warning_captured at: _pytest.config.__init__.py:585
+        """
+        testdir.makepyfile(
+            some_plugin="""
+            import pytest
+            pytest.skip("thing", allow_module_level=True)
+            """
+        )
+        result = testdir.runpytest("-p", "some_plugin", syspathinsert=True)
+        # with stacklevel=2 the warning originates from
+        # config.PytestPluginManager.import_plugin  is thrown by a skipped plugin
+        result.stdout.fnmatch_lines(
+            [
+                "*config{sep}__init__.py*".format(sep=os.sep),
+                "*skipped plugin 'some_plugin': thing*",
+            ]
+        )
+
+    def test_issue4445_resultlog(self, testdir):
+        """#4445: Make sure the warning points to a reasonable location
+        See origin of _issue_warning_captured at: _pytest.resultlog.py:35
+        """
+        testdir.makepyfile(
+            """
+            def test_dummy():
+                pass
+        """
+        )
+        result = testdir.runpytest(
+            "--result-log={dir}".format(dir=testdir.tmpdir.join("result.log"))
+        )
+        # with stacklevel=2 the warning originates from resultlog.pytest_configure
+        # and is thrown when --result-log is used
+        result.stdout.fnmatch_lines(["*resultlog.py*", "*--result-log is deprecated*"])
+
+    def test_issue4445_cacheprovider_set(self, testdir):
+        """#4445: Make sure the warning points to a reasonable location
+        See origin of _issue_warning_captured at: _pytest.cacheprovider.py:59
+        """
+        testdir.tmpdir.join(".pytest_cache").write("something wrong")
+        result = testdir.runpytest()
+        # with stacklevel=3 the warning originates from one stacklevel above
+        # _issue_warning_captured in cacheprovider.Cache.set and is thrown
+        # when there are errors during cache folder creation
+        result.stdout.fnmatch_lines(
+            ["*cacheprovider.py*", "*could not create cache path*"]
+        )
+
+    def test_issue4445_issue5928_mark_generator(self, testdir, request):
+        """#4445 and #5928: Make sure the warning from an unknown mark points to
+        the test file where this mark is used.
+        """
+        testdir.makepyfile(
+            """
+            import pytest
+
+            @pytest.mark.unknown
+            def test_it():
+                pass
+            """
+        )
+        result = testdir.runpytest_subprocess()
+        # with stacklevel=2 the warning should originate from the above created test file
+        # referenced by request.node.name
+        result.stdout.fnmatch_lines(
+            [
+                "*{testfile}.py:3*".format(testfile=request.node.name),
```

Comment:
You could just use `str(p1)` here with `p1 = testdir.makepyfile…` above (I guess).

Reply:
The test should be more specific (hardened) though, checking for the warning header above, and could match the whole line etc.
Like it is now it would still pass if theres an INTERNALERROR where the file is mentioned.
Also `runpytest_subprocess` should not be necessary, is it?

Reply:
Also matching the "*= 1 passed, 1 warning in *" line is recommended (ensures also the exitcode).

Reply:
> You could just use `str(p1)` here with `p1 = testdir.makepyfile…` above (I guess).

I will have a look at this, as this surely can be simpler

>The test should be more specific (hardened) though, checking for the warning header above, > and could match the whole line etc.

I agree totally that they should be more strict, but also thought the most essential parts are the locations, because of the changed stacklevel. And for determining the exact location we are limited now, because of the implications regarding the usage of frame. For me i doesn't make sense to match against more text. I will think about it.

> Also `runpytest_subprocess` should not be necessary, is it?

There was a reason for using `runpytest_subprocess`: Using only `runpytest` resulted in no warning at all for that specific test.

> Also matching the "*= 1 passed, 1 warning in *" line is recommended (ensures also the exitcode). 

Good point!